File: cpp/atbash-cipher/atbash_cipher.cpp
#include "atbash_cipher.h"

using namespace std;

namespace atbash_cipher {

bool is_char(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

bool is_digit(char c) {
    return c >= '0' && c <= '9';
}

char encode_char(char c) {
    return 'z' - tolower(c) + 'a';
}

char decode_char(char c) {
    return ((c - 'z') * -1) + 'a';
}

string encode(string input) {
    const int block_length = 5;
    string encoded = "";
    int encoded_chars = 0;

    for (unsigned int i = 0; i < input.length(); ++i) {
        char encoded_char; 

        if (is_char(input.at(i))) {
            encoded_char = encode_char(input.at(i));
        } else if (is_digit(input.at(i))) {
            encoded_char = input.at(i);
        } else {
            continue;
        }

        if (encoded_chars > 0 && encoded_chars % block_length == 0) {
            encoded.push_back(' ');
        }

        encoded_chars++;
        encoded.push_back(encoded_char);
    }

    return encoded;
}

string decode(string encoded) {
    string decoded = "";

    for (unsigned int i = 0; i < encoded.length(); ++i) {
        if (encoded.at(i) == ' ') {
            continue;
        }

        char decoded_char = is_char(encoded.at(i)) ? decode_char(encoded.at(i)) : encoded.at(i);

        decoded.push_back(decoded_char);
    }

    return decoded;
}

}  // namespace atbash_cipher

File: lua/atbash-cipher/atbash-cipher.lua
local char_mappings = {
  a = 'z',
  b = 'y',
  c = 'x',
  d = 'w',
  e = 'v',
  f = 'u',
  g = 't',
  h = 's',
  i = 'r',
  j = 'q',
  k = 'p',
  l = 'o',
  m = 'n',
  n = 'm',
  o = 'l',
  p = 'k',
  q = 'j',
  r = 'i',
  s = 'h',
  t = 'g',
  u = 'f',
  v = 'e',
  w = 'd',
  x = 'c',
  y = 'b',
  z = 'a',
}

local function is_char(c)
  return c >= 'a' and c <= 'z' or c >= 'A' and c <= 'Z'
end

local function is_digit(c)
  return c >= '0' and c <= '9'
end

return {
  encode = function(plaintext)
    local encoded = ""
    local j = 0

    for i = 1, #plaintext do
      local c = string.sub(plaintext, i, i)

      if (is_char(c) or is_digit(c)) and j > 0 and j % 5 == 0 then
        encoded = encoded .. " "
      end      

      if is_char(c) then
        encoded = encoded .. char_mappings[string.lower(c)]
        j = j + 1
      elseif is_digit(c) then
        encoded = encoded .. c
        j = j + 1
      end
    end

    return encoded
  end
}
